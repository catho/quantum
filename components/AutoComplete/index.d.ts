import { FC } from 'react';
export interface AutoCompleteProps {
  suggestions: Array<string>;
  id?: string;
  name?: string;
  label?: string;
  value?: string;
  helperText?: string;
  error?: string;
  disabled?: boolean;
  minlength?: string;
  placeholder?: string;
  onChange?: (value: string) => void;
  onSelectedItem?: (item: string) => void;
}

declare const AutoComplete: FC<AutoCompleteProps>;
export default AutoComplete;