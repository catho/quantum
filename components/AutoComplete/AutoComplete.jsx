import { useState, useRef, useEffect } from 'react';
import PropTypes from 'prop-types';
import styled, { css } from 'styled-components';
import { shadow, normalizeChars } from '../shared';
import { baseFontSize, colors, spacing } from '../shared/theme';
import Icon from '../Icon';

import {
  TextInput,
  InputLabel,
  InputErrorMessage,
  HelperText,
  RequiredMark,
} from '../Input/sub-components';
import useKeyPress from './SubComponents/UseKeyPress';

const ITEM_HEIGHT = '44px';
const MAX_ITEMS_VISIBILITY = 7;
const DROPITEM_FONT_SIZE = baseFontSize * 0.875;

const ComponentWrapper = styled.div`
  ${({
    theme: {
      colors: { neutral },
    },
    skin = 'default',
  }) => css`
    position: relative;
    color: ${skin === 'default' ? neutral[700] : neutral[0]};
  `}
`;

const InputWrapper = styled.div`
  position: relative;
`;

const InputText = styled(TextInput)`
  ${({
    theme: {
      spacing: { xsmall },
    },
  }) => css`
    margin-top: ${xsmall}px;
  `}
`;

const InputIcon = styled(Icon)`
  cursor: pointer;
  position: absolute;
  ${({
    theme: {
      spacing: { xsmall, medium },
    },
  }) => css`
    right: ${medium}px;
    bottom: ${xsmall * 1.25}px;
    width: ${baseFontSize * 1.5}px;
  `}
`;

const InputErrorIcon = styled(InputIcon).attrs({ name: 'error' })`
  ${({
    theme: {
      colors: {
        error: { 700: error700 },
      },
    },
    skin,
  }) => css`
    color: ${skin === 'default' ? error700 : 'inherit'};
  `}
`;

const List = styled.ul`
  border-radius: 4px;
  box-sizing: border-box;
  list-style: none;
  max-height: calc(${ITEM_HEIGHT} * ${MAX_ITEMS_VISIBILITY});
  overflow: auto;
  padding: 0;
  position: absolute;
  width: 100%;
  z-index: 1;

  ${({ theme }) => {
    const {
      spacing: { xxsmall },
      colors: {
        neutral: { 0: neutral0, 700: neutral700 },
      },
    } = theme;
    return css`
      background-color: ${neutral0};
      margin-top: ${xxsmall}px;
      ${shadow(3, neutral700)({ theme })};
    `;
  }}
`;

const ListItem = styled.li`
  display: flex;
  align-items: center;
  justify-content: space-between;
  box-sizing: border-box;
  cursor: pointer;
  height: ${ITEM_HEIGHT};
  ${({
    theme: {
      spacing: { xsmall, medium },
      colors: {
        neutral: { 0: neutral0, 700: neutral700 },
      },
    },
  }) => css`
    color: ${neutral700};
    font-size: ${DROPITEM_FONT_SIZE}px;
    background-color: ${neutral0};
    padding: ${xsmall}px ${medium}px;
  `}

  &[aria-selected = 'true' ], &:hover {
    ${({
      theme: {
        colors: {
          neutral: { 100: neutral100 },
        },
      },
    }) => css`
      background-color: ${neutral100};
    `}
  }
`;

const PoliteStatus = styled.div`
  border: 0;
  clip: rect(0 0 0 0);
  height: 1px;
  margin: -1px;
  overflow: hidden;
  padding: 0;
  position: absolute;
  width: 1px;
`;

const AutoComplete = ({
  id,
  name,
  label,
  value,
  error,
  disabled,
  helperText,
  placeholder,
  suggestions,
  theme,
  onSelectedItem,
  onChange,
  required,
  skin,
}) => {
  const [userTypedValue, setUserTypedValue] = useState(value);
  const [filterSuggestions, setFilterSuggestions] = useState(suggestions);
  const [filterSuggestionsLength, setFilterSuggestionsLength] = useState(
    filterSuggestions.length,
  );
  const [showSuggestions, setShowSuggestions] = useState(false);
  const [cursor, setCursor] = useState(0);
  const wrapperRef = useRef();
  const listOptions = useRef();
  const autoInputRef = useRef(null);
  const EscapeKeyPressValue = 'Escape';
  const assistiveDescriptionDefault = `Digite uma ou mais letras para expandir os resultados. ${filterSuggestionsLength} estão disponiveis.`;
  const assistiveDescriptionDropDownOpen = `${assistiveDescriptionDefault} ${
    filterSuggestions[cursor]
  } 
  ${cursor + 1} de ${filterSuggestionsLength} está destacado`;

  const filterItems = (currentValue) =>
    suggestions.filter((suggestion) => {
      let option = normalizeChars(suggestion.toLowerCase());
      option = normalizeChars(option);
      return option.indexOf(normalizeChars(currentValue.toLowerCase())) > -1;
    });

  const handleFilter = (currentValue) => {
    const filteredValues = filterItems(currentValue);
    setShowSuggestions(!!filteredValues.length);
    setFilterSuggestions(filteredValues);
  };

  const handleChange = (currentValue) => {
    setUserTypedValue(currentValue);
    onChange(currentValue);
    setCursor(0);
    handleFilter(currentValue);
  };

  const handleClickOutside = (event) => {
    if (!wrapperRef.current?.contains(event.target)) {
      setShowSuggestions(false);
    }
  };

  const handleInputClick = () => {
    const filteredItem = filterItems(userTypedValue);
    setFilterSuggestions(filteredItem);
    setShowSuggestions(true);
  };

  const handleItemClick = (item) => {
    setUserTypedValue(item);
    onSelectedItem(item);
    setShowSuggestions(false);
  };

  const handleClearValue = () => {
    setUserTypedValue('');
    onChange('');
    setFilterSuggestions(suggestions);
    setCursor(0);
  };

  const handleEscPress = ({ key }) => {
    const node = wrapperRef.current;
    if (node && key === EscapeKeyPressValue) {
      setShowSuggestions(false);
      node.children[1].focus();
      setCursor(0);
    }
  };

  const generateSuggestions = () =>
    showSuggestions && filterSuggestions ? (
      <List
        theme={theme}
        role="listbox"
        id="autocompleteOptions"
        ref={listOptions}
        skin={skin}
      >
        {filterSuggestions.map((item, index) => (
          <ListItem
            key={item}
            aria-posinset={index}
            onClick={() => handleItemClick(item)}
            theme={theme}
            aria-selected={index === cursor}
            role="option"
            tabIndex="-1"
          >
            {item}
          </ListItem>
        ))}
      </List>
    ) : null;

  const generateAssistiveDescript = () => {
    if (showSuggestions) {
      return assistiveDescriptionDropDownOpen;
    }
    return assistiveDescriptionDefault;
  };

  const downPress = useKeyPress('ArrowDown');
  const upPress = useKeyPress('ArrowUp');
  const enterPress = useKeyPress('Enter');
  const tabPress = useKeyPress('Tab');

  useEffect(() => {
    setFilterSuggestionsLength(filterSuggestions?.length);
  }, [filterSuggestions]);

  /* istanbul ignore next */
  useEffect(() => {
    if (showSuggestions && filterSuggestionsLength && downPress) {
      const selectedCursor =
        cursor < filterSuggestionsLength - 1 ? cursor + 1 : cursor;
      setCursor(selectedCursor);
      listOptions.current.children[selectedCursor].focus();
    }
  }, [downPress]);

  /* istanbul ignore next */
  useEffect(() => {
    if (showSuggestions && filterSuggestionsLength && upPress && cursor > 0) {
      const selectedCursor = cursor > 0 ? cursor - 1 : cursor;
      setCursor(selectedCursor);
      listOptions.current.children[selectedCursor].focus();
    }
    if (upPress && cursor === 0) {
      wrapperRef.current.children[1].focus();
      setCursor(-1);
    }
  }, [upPress]);

  /* istanbul ignore next */
  useEffect(() => {
    if (showSuggestions && filterSuggestionsLength && enterPress) {
      setUserTypedValue(filterSuggestions[cursor]);
      onSelectedItem(filterSuggestions[cursor]);
      setShowSuggestions(false);
      wrapperRef.current.children[1].focus();
    }
  }, [cursor, enterPress]);

  /* istanbul ignore next */
  useEffect(() => {
    if (showSuggestions && tabPress) {
      setUserTypedValue(filterSuggestions[cursor]);
      onSelectedItem(filterSuggestions[cursor]);
      setShowSuggestions(false);
    } else if (tabPress) {
      setShowSuggestions(false);
    }
  }, [tabPress]);

  useEffect(() => {
    window.addEventListener('click', handleClickOutside);
    window.addEventListener('keydown', handleEscPress);
    return () => {
      window.removeEventListener('click', handleClickOutside);
      window.removeEventListener('keydown', handleEscPress);
    };
  }, []);

  useEffect(() => {
    if (document.activeElement === autoInputRef.current) {
      handleFilter(userTypedValue);
    }
  }, [suggestions]);

  return (
    <ComponentWrapper theme={theme} skin={skin}>
      <InputWrapper ref={wrapperRef}>
        <InputLabel htmlFor={id} error={error}>
          {label}
          {required && <RequiredMark skin={skin}>*</RequiredMark>}
        </InputLabel>
        <InputText
          id={id}
          ref={autoInputRef}
          name={name}
          type="text"
          error={error}
          disabled={disabled}
          placeholder={placeholder}
          autoComplete="off"
          role="combobox"
          aria-owns="autocompleteOptions"
          aria-autocomplete="both"
          aria-expanded={showSuggestions}
          value={userTypedValue}
          onClick={() => handleInputClick()}
          onChange={(e) => handleChange(e.target.value)}
          skin={skin}
          required={required}
        />
        {userTypedValue && !error && !disabled && (
          <InputIcon
            theme={theme}
            name="clear"
            description="limpar valor"
            onClick={() => handleClearValue()}
          />
        )}
        {generateSuggestions()}
        {error && (
          <InputErrorIcon description={error} theme={theme} skin={skin} />
        )}
      </InputWrapper>
      {helperText && <HelperText>{helperText}</HelperText>}
      {error && (
        <InputErrorMessage theme={theme} skin={skin}>
          {error}
        </InputErrorMessage>
      )}
      <PoliteStatus role="status" aria-atomic="true" aria-live="polite">
        {generateAssistiveDescript()}
      </PoliteStatus>
    </ComponentWrapper>
  );
};

AutoComplete.propTypes = {
  theme: PropTypes.shape({
    colors: PropTypes.object,
    spacing: PropTypes.object,
  }),
  /** A list of string or objects with the values to show in component */
  suggestions: PropTypes.arrayOf(PropTypes.string).isRequired,
  id: PropTypes.string,
  /** Displays an error message and changes border color to error color */
  error: PropTypes.string,
  disabled: PropTypes.bool,
  /** Displays a label text that describes the field */
  label: PropTypes.string,
  value: PropTypes.string,
  name: PropTypes.string,
  placeholder: PropTypes.string,
  /** Callback function to receive what the user is typing */
  onChange: PropTypes.func,
  /** Callback function to receive user selected value */
  onSelectedItem: PropTypes.func,
  /** Displays a helper text below the component */
  helperText: PropTypes.string,
  required: PropTypes.bool,
  skin: PropTypes.oneOf(['default', 'dark']),
};

AutoComplete.defaultProps = {
  theme: {
    spacing,
    colors,
  },
  id: '',
  name: '',
  label: '',
  value: '',
  helperText: '',
  error: '',
  disabled: false,
  placeholder: 'Select an option',
  onChange: () => {},
  onSelectedItem: () => {},
  required: false,
  skin: 'default',
};

export default AutoComplete;
