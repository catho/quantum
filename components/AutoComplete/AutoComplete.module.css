:root {
    --max-items-visibility-autocomplete: 7;
    --item-height-autocomplete: 44px;
    --dropitem-font-size: calc(var(--qtm-base-font-size) * 0.875);
}

.wrapper {
    position: relative;
    color: var(--qtm-colors-neutral-700);
}

.input-wrapper {
    position: relative;
}

.wrapper-dark {
    color: var(--qtm-colors-neutral-0);
}

.input-text {
    margin-top: var(--qtm-spacing-xsmall);
}

.input-icon {
    cursor: pointer;
    position: absolute;
    right: var(--qtm-spacing-medium);
    bottom: calc(var(--qtm-spacing-xsmall) * 1.25);
    width: calc(var(--qtm-base-font-size) * 1.5);
}

.input-error-icon-default {
    color: var(--qtm-colors-error-700);
}

.input-error-icon-dark {
    color: inherit;
}

.polite-status {
    border: 0;
    clip: rect(0 0 0 0);
    height: 1px;
    margin: -1px;
    overflow: hidden;
    padding: 0;
    position: absolute;
    width: 1px;
}

.list-suggestions {
    border-radius: 4px;
    box-sizing: border-box;
    list-style: none;
    max-height: calc(var(--item-height-autocomplete) * var(--max-items-visibility-autocomplete));
    overflow: auto;
    padding: 0;
    position: absolute;
    width: 100%;
    z-index: 1;
    background-color: var(--qtm-colors-neutral-0);
    margin-top: var(--qtm-spacing-xxsmall);
}

.item-suggestions {
    display: flex;
    align-items: center;
    justify-content: space-between;
    box-sizing: border-box;
    cursor: pointer;
    height: var(--item-height-autocomplete);
    color: var(--qtm-colors-neutral-700);
    font-size: var(--dropitem-font-size);
    background-color: var(--qtm-colors-neutral-0);
    padding: var(--qtm-spacing-xsmall) var(--qtm-spacing-medium);
}

.item-suggestions[aria-selected='true'],
.item-suggestions:hover {
    background-color: var(--qtm-colors-neutral-100);
}