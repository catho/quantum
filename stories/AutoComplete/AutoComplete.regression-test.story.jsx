import { AutoComplete } from '../../components';

const Examples = ['morango', 'melancia', 'maÃ§a', 'banana', 'laranja'];

export default {
  title: 'AutoComplete',
  component: AutoComplete,
};

const Template = (args) => <AutoComplete {...args} />;

export const Default = Template.bind({});
Default.args = {
  id: 'autoCompleteNew',
  name: 'autoCompleteNew',
  label: 'Auto Complete',
  suggestions: Examples,
  required: true,
};

export const WithError = Template.bind({});
WithError.args = {
  ...Default.args,
  error: 'Error message',
};

export const WithHelperText = Template.bind({});
WithHelperText.args = {
  ...Default.args,
  helperText: 'Helper text',
};

export const WithDark = Template.bind({});
WithDark.args = {
  ...Default.args,
  skin: 'dark',
  error: 'Error message',
  helperText: 'Helper text',
};

WithDark.parameters = {
  backgrounds: {
    default: 'dark',
    values: [{ name: 'dark', value: '#1250C4' }],
  },
};
