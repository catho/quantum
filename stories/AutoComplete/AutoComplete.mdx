import { Meta, Canvas, Story, ArgsTable } from '@storybook/addon-docs';
import { Header } from '../shared';
import AutoComplete from '../../components/AutoComplete';
import { Default, Required, Disabled, WithHelperText, WithLabel, WithError, WithSkin } from './AutoComplete.stories.jsx';

<Meta title="Forms/AutoComplete" component={AutoComplete}  />

<Header title="AutoComplete">
    The Autocomplete is an input that shows a dropdown menu with suggestions according to the written text.
</Header>

- [Importing and usage](#importing)
- [API](#api)
- [Examples](#examples)

## <a id="importing"></a>Importing and usage

```js
import AutoComplete from '@catho/quantum/AutoComplete';
```

<Canvas>
  <Story of={Default}/>
</Canvas>

## <a id="api"></a>API

<ArgsTable of={AutoComplete} />

## <a id="examples"></a>Examples

### Required

<Canvas>
  <Story of={Required}/>
</Canvas>

### Disabled

<Canvas>
  <Story of={Disabled}/>
</Canvas>

### With Helper Text

<Canvas>
  <Story of={WithHelperText}/>
</Canvas>

### With Label

<Canvas>
  <Story of={WithLabel}/>
</Canvas>

### With Error

<Canvas>
  <Story of={WithError}/>
</Canvas>

### With Skin

<Canvas>
  <Story of={WithSkin}/>
</Canvas>
