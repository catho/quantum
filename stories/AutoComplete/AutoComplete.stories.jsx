import AutoComplete from '../../components/AutoComplete';

export default {
  title: 'Forms/AutoComplete',
  component: AutoComplete,
};

const itemsMock = [
  'Analista de Gestão de Incidentes',
  'Analista de Governança de TI',
  'Analista de Help Desk',
  'Analista de Negócios de TI',
  'Analista de Garantia da Qualidade',
  'Analista de Pricing Sênior',
  'Analista de Pricing Pleno',
  'Analista de Pricing Junior',
  'Analista de Engenharia de Produto',
  'Analista de Sustentabilidade',
  'Cronoanalista',
];

const Template = args => <AutoComplete suggestions={itemsMock} {...args} />;

export const Default = Template.bind({});

export const Required = Template.bind({});
Required.args = {
  label: 'Some label...',
  required: true,
};

export const Disabled = Template.bind({});
Disabled.args = {
  disabled: true,
};

export const WithHelperText = Template.bind({});
WithHelperText.args = {
  helperText: 'Some text...',
};

export const WithError = Template.bind({});
WithError.args = {
  error: 'Some error text',
};

export const WithLabel = Template.bind({});
WithLabel.args = {
  label: 'Some text label...',
};

export const WithSkin = Template.bind({});
WithSkin.args = {
  skin: 'dark',
};
