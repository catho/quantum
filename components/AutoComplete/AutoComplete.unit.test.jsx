import { render, screen, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import AutoComplete from './AutoComplete';

const Examples = ['morango', 'melancia', 'maça', 'banana', 'laranja'];
const ExamplesChanged = ['melanci', 'melancia'];

const ARROW_DOWN_KEY_CODE = '{ArrowDown}';
const ARROW_UP_KEY_CODE = '{ArrowUp}';
const ENTER_KEY_CODE = '{Enter}';
const ESCAPE_KEY_CODE = '{Escape}';

describe('AutoComplete', () => {
  it('Should render Autocomplete', () => {
    const { container } = render(<AutoComplete suggestions={Examples} />);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('Should render Autocomplete with dark skin', () => {
    const { container } = render(
      <AutoComplete skin="dark" suggestions={Examples} />,
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it('should render the AutoComplete open', async () => {
    render(<AutoComplete suggestions={Examples} />);
    const input = screen.getByRole('combobox');

    await userEvent.type(input, 'm');

    const autoCompleteOptions = (await screen.findAllByRole('listbox'))[0];

    expect(autoCompleteOptions).toBeInTheDocument();
  });

  it('should open the options when input is clicked', async () => {
    render(<AutoComplete suggestions={Examples} />);
    const input = screen.getByRole('combobox');

    await userEvent.click(input);

    const autoCompleteOptions = (await screen.findAllByRole('listbox'))[0];

    expect(autoCompleteOptions).toBeInTheDocument();
  });

  it('should select a item when is clicked', async () => {
    render(<AutoComplete suggestions={Examples} />);
    const input = screen.getByRole('combobox');

    await userEvent.type(input, 'melanci');

    const melanciaOption = (
      await screen.findAllByRole('option', {
        name: /melancia/i,
      })
    )[0];

    await userEvent.click(melanciaOption);

    expect(input.getAttribute('value')).toBe('melancia');
  });

  it('should close the options after item is clicked', async () => {
    render(<AutoComplete suggestions={Examples} />);
    const input = screen.getByRole('combobox');

    await userEvent.type(input, 'melanci');

    const autoCompleteOptions = (await screen.findAllByRole('listbox'))[0];

    expect(autoCompleteOptions).toBeInTheDocument();

    const melanciaOption = (
      await screen.findAllByRole('option', {
        name: /melancia/i,
      })
    )[0];

    await userEvent.click(melanciaOption);

    expect(autoCompleteOptions).not.toBeInTheDocument();
  });

  it('should correctly filter items with accent', async () => {
    render(
      <AutoComplete
        suggestions={['morango', 'melancia', 'maça', 'banana', 'laranja']}
      />,
    );
    const input = screen.getByRole('combobox');

    await userEvent.type(input, 'maçã');

    const maçaOption = (
      await screen.findAllByRole('option', {
        name: /maça/i,
      })
    )[0];

    expect(maçaOption).toBeInTheDocument();
  });

  it('should correctly filter items without accent', async () => {
    render(
      <AutoComplete
        suggestions={['morango', 'melancia', 'maçã', 'banana', 'laranja']}
      />,
    );
    const input = screen.getByRole('combobox');

    await userEvent.type(input, 'maça');

    const maçaOption = (
      await screen.findAllByRole('option', {
        name: /maçã/i,
      })
    )[0];

    expect(maçaOption).toBeInTheDocument();
  });

  it('should close the options when there is no valid option', async () => {
    render(<AutoComplete suggestions={Examples} />);

    const input = screen.getByRole('combobox');

    await userEvent.click(input);

    const autoCompleteOptions = screen.getAllByRole('listbox')[0];

    await userEvent.type(input, 'mon');

    expect(autoCompleteOptions).not.toBeInTheDocument();
  });

  it('should clean input value when clicked in the x button', async () => {
    const { container } = render(<AutoComplete suggestions={Examples} />);

    const input = screen.getByRole('combobox');

    await userEvent.type(input, 'mon');

    const cleanButton = container.querySelectorAll('svg')[0];

    await userEvent.click(cleanButton);

    expect(input.getAttribute('value')).toBe('');
  });

  it('should has an error icon when "error" prop is set', () => {
    const { container } = render(
      <AutoComplete suggestions={Examples} error="error text" />,
    );

    const InputErrorIconElement = container.querySelectorAll('svg')[0];

    expect(InputErrorIconElement).toBeInTheDocument();
  });

  it('should call onChangeMock callback when prop is setted', async () => {
    const onChangeMock = jest.fn();
    render(<AutoComplete suggestions={Examples} onChange={onChangeMock} />);

    const input = screen.getByRole('combobox');

    await userEvent.type(input, 'mora');

    expect(onChangeMock).toHaveBeenCalled();
  });

  it('should call selectedItemMock callback when prop is setted and user selected a option', async () => {
    const selectedItemMock = jest.fn();
    render(
      <AutoComplete suggestions={Examples} onSelectedItem={selectedItemMock} />,
    );

    const input = screen.getByRole('combobox');

    await userEvent.type(input, 'melanci');

    const autoCompleteOptions = screen.getByRole('option', {
      name: /melancia/i,
    });

    await userEvent.click(autoCompleteOptions);

    expect(selectedItemMock).toHaveBeenCalled();
  });

  it('should has an helper text prop is set', () => {
    render(<AutoComplete suggestions={Examples} helperText="helper text" />);

    expect(screen.getByText(/helper text/i)).toBeInTheDocument();
  });

  it('should filter the options if suggestions prop is changed', async () => {
    const { rerender } = render(<AutoComplete suggestions={Examples} />);

    const input = screen.getByRole('combobox');

    userEvent.type(input, 'melanci');

    rerender(<AutoComplete suggestions={ExamplesChanged} />);

    // Esse teste está correto? Na doc foram feitos testes e o valor digitado no input
    // Não vira um dos itens da lista da sugestão. Por exemplo, somente
    // "melancia" é exibido e não "melanc" e "melancia"

    await waitFor(() => expect(input).toHaveValue('melanci'));
    expect(await screen.findAllByRole('option')).toHaveLength(2);
    expect(screen.getByRole('option', { name: 'melanci' })).toBeInTheDocument();
    expect(
      screen.getByRole('option', { name: 'melancia' }),
    ).toBeInTheDocument();
  });

  it('Should have a value when prop value is set', () => {
    const initialValue = 'initial value';
    render(<AutoComplete value={initialValue} suggestions={[]} />);
    const autoCompleteInput = screen.getByRole('combobox');
    expect(autoCompleteInput).toHaveAttribute('value', initialValue);
  });

  it('Should close options when user press Escape', async () => {
    render(<AutoComplete suggestions={Examples} />);
    const input = screen.getByRole('combobox');

    await userEvent.type(input, 'moran');
    const autoCompleteOptionsList = await screen.findByRole('listbox');

    expect(autoCompleteOptionsList).toBeInTheDocument();

    await userEvent.keyboard(ESCAPE_KEY_CODE);

    expect(autoCompleteOptionsList).not.toBeInTheDocument();
  });

  it('Should allow User to select an option using only keyboard', async () => {
    const onSelectedItemMock = jest.fn();
    render(
      <AutoComplete
        suggestions={Examples}
        onSelectedItem={onSelectedItemMock}
      />,
    );
    const input = screen.getByRole('combobox');

    await userEvent.type(input, 'm');
    const autoCompleteOptionsList = await screen.findByRole('listbox');

    expect(autoCompleteOptionsList).toBeInTheDocument();

    await userEvent.keyboard(
      `${ARROW_DOWN_KEY_CODE}${ARROW_DOWN_KEY_CODE}${ARROW_UP_KEY_CODE}`,
    );
    const autoCompleteOptionItem = screen.getByRole('option', {
      name: /morango/i,
    });

    expect(autoCompleteOptionItem).toHaveFocus();
    await userEvent.keyboard(ENTER_KEY_CODE);

    expect(input.value).toEqual('morango');
    expect(onSelectedItemMock).toHaveBeenCalledWith('morango');
    expect(autoCompleteOptionsList).not.toBeInTheDocument();
  });

  it('Should focus on input when there are no items to navigate above', async () => {
    const onSelectedItemMock = jest.fn();
    render(
      <AutoComplete
        suggestions={Examples}
        onSelectedItem={onSelectedItemMock}
      />,
    );
    const input = screen.getByRole('combobox');

    await userEvent.type(input, 'm');
    const autoCompleteOptionsList = await screen.findByRole('listbox');

    expect(autoCompleteOptionsList).toBeInTheDocument();

    await userEvent.keyboard(`${ARROW_DOWN_KEY_CODE}${ARROW_UP_KEY_CODE}`);
    expect(input).toHaveFocus();
  });
});
