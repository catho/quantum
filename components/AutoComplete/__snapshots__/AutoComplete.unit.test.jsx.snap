// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AutoComplete Should render Autocomplete 1`] = `
.InputLabel-module__input-label___u-ovg {
  color: inherit;
  display: block;
  font-weight: 700;
  margin: var(--qtm-spacing-xsmall) 0 0;
}

.TextInput-module__text-input___VCylJ {
  -webkit-appearance: none;
  background-color: var(--qtm-colors-neutral-0);
  border: 2px solid var(--qtm-colors-neutral-500);
  border-radius: 4px;
  box-sizing: border-box;
  color: var(--qtm-colors-neutral-700);
  font-size: medium;
  letter-spacing: .2px;
  margin-top: 0;
  outline: none;
  padding: var(--qtm-spacing-xsmall) var(--qtm-spacing-medium);
  transition: all .2s ease-in-out;
  width: 100%;
}

.c0 {
  position: relative;
  color: #424242;
}

.c1 {
  position: relative;
}

.c2 {
  margin-top: 8px;
}

.c3 {
  border: 0;
  -webkit-clip: rect(0 0 0 0);
  clip: rect(0 0 0 0);
  height: 1px;
  margin: -1px;
  overflow: hidden;
  padding: 0;
  position: absolute;
  width: 1px;
}

<div
  class="c0"
>
  <div
    class="c1"
  >
    <label
      class="InputLabel-module__input-label___u-ovg"
      for=""
    />
    <input
      aria-autocomplete="both"
      aria-expanded="false"
      aria-owns="autocompleteOptions"
      autocomplete="off"
      class="TextInput-module__text-input___VCylJ c2"
      id=""
      name=""
      placeholder="Select an option"
      role="combobox"
      style="--icon-size: var(--qtm-spacing-large); --icon-area: calc(var(--qtm-spacing-medium) + var(--icon-size) + var(--qtm-spacing-xsmall));"
      type="text"
      value=""
    />
  </div>
  <div
    aria-atomic="true"
    aria-live="polite"
    class="c3"
    role="status"
  >
    Digite uma ou mais letras para expandir os resultados. 5 estão disponiveis.
  </div>
</div>
`;

exports[`AutoComplete Should render Autocomplete with dark skin 1`] = `
.InputLabel-module__input-label___u-ovg {
  color: inherit;
  display: block;
  font-weight: 700;
  margin: var(--qtm-spacing-xsmall) 0 0;
}

.TextInput-module__text-input___VCylJ {
  -webkit-appearance: none;
  background-color: var(--qtm-colors-neutral-0);
  border: 2px solid var(--qtm-colors-neutral-500);
  border-radius: 4px;
  box-sizing: border-box;
  color: var(--qtm-colors-neutral-700);
  font-size: medium;
  letter-spacing: .2px;
  margin-top: 0;
  outline: none;
  padding: var(--qtm-spacing-xsmall) var(--qtm-spacing-medium);
  transition: all .2s ease-in-out;
  width: 100%;
}

.TextInput-module__text-input-dark___TrWzt {
  background-color: rgba(0,0,0,.7)!important;
  border: 2px solid transparent;
  color: var(--qtm-colors-neutral-0);
}

.c0 {
  position: relative;
  color: #ffffff;
}

.c1 {
  position: relative;
}

.c2 {
  margin-top: 8px;
}

.c3 {
  border: 0;
  -webkit-clip: rect(0 0 0 0);
  clip: rect(0 0 0 0);
  height: 1px;
  margin: -1px;
  overflow: hidden;
  padding: 0;
  position: absolute;
  width: 1px;
}

<div
  class="c0"
>
  <div
    class="c1"
  >
    <label
      class="InputLabel-module__input-label___u-ovg"
      for=""
    />
    <input
      aria-autocomplete="both"
      aria-expanded="false"
      aria-owns="autocompleteOptions"
      autocomplete="off"
      class="TextInput-module__text-input___VCylJ TextInput-module__text-input-dark___TrWzt c2"
      id=""
      name=""
      placeholder="Select an option"
      role="combobox"
      style="--icon-size: var(--qtm-spacing-large); --icon-area: calc(var(--qtm-spacing-medium) + var(--icon-size) + var(--qtm-spacing-xsmall));"
      type="text"
      value=""
    />
  </div>
  <div
    aria-atomic="true"
    aria-live="polite"
    class="c3"
    role="status"
  >
    Digite uma ou mais letras para expandir os resultados. 5 estão disponiveis.
  </div>
</div>
`;
